source/opt/optimizer.hpp

```diff
@@ -0,0 +1,151 @@
+
+  // Registers a fold-spec-constant-op-and-composite pass to this optimizer.
+  // A fold-spec-constant-op-and-composite pass folds spec constants defined by
+  // OpSpecConstantOp or OpSpecConstantComposite instruction, to normal
+  // Constants defined by OpConstantTrue, OpConstantFalse, OpConstant,
+  // OpConstantNull, or OpConstantComposite instructions. Note that spec
+  // constants defined with OpSpecConstant, OpSpecConstantTrue, or
+  // OpSpecConstantFalse instructions are not handled, as these instructions
+  // indicate their value are not determined and can be changed in future.
+  // A spec constant is foldable if all of its value(s) can be determined from
+  // the module. E.g., an integer spec constant defined with OpSpecConstantOp
+  // instruction can be folded if its value won't change later. This pass will
+  // replace the original OpSpecContantOp instruction with an OpConstant
+  // instruction. When folding composite spec constants, new instructions
+  // may be inserted to define the components of the composite constant first,
+  // then the original spec constants will be replaced by OpConstantComposite
+  // instructions.
+  //
+  // There are some operations not supported yet:
+  //   OpSConvert, OpFConvert, OpQuantizeToF16 and
+  //   all the operations under Kernel capability.
+  // TODO(qining): Add support for the operations listed above.
+  Optimizer& RegisterFoldSpecConstantOpAndCompositePass();
+
+  // Registers a unify-constant pass to this optimizer.
+  // A unify-constant pass de-duplicates the constants. Constants with the exact
+  // same value and identical form will be unified and only one constant will
+  // be kept for each unique pair of type and value.
+  // There are several cases not handled by this pass:
+  //  1) Constants defined by OpConstantNull instructions (null constants) and
+  //  constants defined by OpConstantFalse, OpConstant or OpConstantComposite
+  //  with value 0 (zero-valued normal constants) are not considered equivalent.
+  //  So null constants won't be used to replace zero-valued normal constants,
+  //  vice versa.
+  //  2) Whenever there are decorations to the constant's result id id, the
+  //  constant won't be handled, which means, it won't be used to replace any
+  //  other constants, neither can other constants replace it.
+  //  3) NaN in float point format with different bit patterns are not unified.
+  Optimizer& RegisterUnifyConstantPass();
+
+  // Registers a eliminate-dead-constant pass to this optimizer.
+  // A eliminate-dead-constant pass removes dead constants, including normal
+  // contants defined by OpConstant, OpConstantComposite, OpConstantTrue, or
+  // OpConstantFalse and spec constants defined by OpSpecConstant,
+  // OpSpecConstantComposite, OpSpecConstantTrue, OpSpecConstantFalse or
+  // OpSpecConstantOp.
+  Optimizer& RegisterEliminateDeadConstantPass();
+
+  // Clears all passes previously registered.
+  Optimizer& ClearRegisteredPasses();
```

Comment:
You don't need this.  Pretty much in any case where you want this you should just make a new optimizer and set its message consumer.